### PR TITLE
[NAE-1468] i18n Data field

### PR DESCRIPTION
**Pull request by camperko**
# Description

Implements [NAE-1468]
- implement i18n field in editable text format

## Dependencies

No new dependencies

### Third party dependencies

No new dependencies were introduced

### Blocking Pull requests

There are no dependencies on other PR

## How Has Been This Tested?

Assisted testing + manual testing + front-end tests

### Test Configuration

| Name                | Tested on |
|---------------------| --------- |
| OS                  | Linux Mint 19.3          |
| Runtime             | node v16.14.2          |
| Dependency Manager  |  npm 8.5.0         |
| Framework version   | Angular v10          |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My changes have been checked, personally or remotely, with @mladoniczky @machacjozef 
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have resolved all conflicts with the target branch of the PR
- [X] I have updated and synced my code with the target branch
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing tests pass locally with my changes:
    - [X] Lint test
    - [X] Unit tests
    - [X] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif)
- [X] I have made corresponding changes to the documentation:
    - [X] Developer documentation
    - [X] User Guides
    - [X] Migration Guides

[NAE-1468]: https://netgrif.atlassian.net/browse/NAE-1468?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by mladoniczky — CHANGES_REQUESTED** [review]

File: projects/netgrif-components-core/src/assets/i18n/de.json [hunk 1]
```diff
@@ -211,7 +211,10 @@
             "weekend": "Das eingegebene Datum muss das Wochenende sein",
             "workday": "Das eingegebene Datum muss ein Wochentag sein",
             "enumeration": "Eine von der angebotenen Optionen muss ausgewählt sein",
-            "min": "Die eingegebene Nummer muss größer als {{length}} sein"
+            "min": "Die eingegebene Nummer muss größer als {{length}} sein",
```
[the remark this answers is not in the document]
Comment (minop) [reply projects/netgrif-components-core/src/assets/i18n/de.json:214]: the translations are correct.

The `translationRequired` and `translationOnly` validation messages are only available in plural forms (for all supported languages), is this intended? Do we want to support singular forms in the future if only one language is required?

The required languages themselves in these messages come from "outside" as parameters. Are the names of these languages also translated?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts [hunk 2]
```diff
@@ -1,27 +1,120 @@
...
+        this.currentValue = I18nField.toObject(newValue);
+        this.refreshFilledMap();
+    }
+
+    public getLanguageIcons() {
+        return this.languageIconsService.languageIcons;
+    }
+
+    public isDefaultValue(choosenLanguage: string) {
+        return choosenLanguage === 'xx';
```
[the remark this answers is not in the document]
Comment (minop) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:75]: why is `'xx'` taken as a placeholder for the defaultValue? Why not `undefined`?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts [hunk 3]
```diff
@@ -4,15 +4,161 @@ import {Layout} from '../../models/layout';
...
+        if (!a && !b) {
+            return true;
+        }
+        if ((!a && !!b) || (!b && !!a)) {
+            return false;
+        }
+        if ((!a.defaultValue && !b.defaultValue) && (!a.translations && !b.translations) && (!a.key && !b.key)) {
+            return true;
+        }
+        if ((!(!a.defaultValue && !b.defaultValue)
```
Comment (minop) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts:62]: could this perhaps be made more readable?

Probably with calls to methods such as `hasTranslations()` and so on

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts [hunk 4]
```diff
@@ -1,21 +1,41 @@
...
+                HttpClientTestingModule
+            ],
+            providers: [
+                {provide: AuthenticationMethodService, useClass: MockAuthenticationMethodService},
+                {provide: ConfigurationService, useClass: TestConfigurationService},
+                {provide: AuthenticationService, useClass: MockAuthenticationService},
+                {provide: UserResourceService, useClass: MockUserResourceService}
+            ],
+            declarations: [I18nDividerFieldComponent, TestWrapperComponent],
+            schemas: [CUSTOM_ELEMENTS_SCHEMA]
```
Comment (minop) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts:37]: why is this necessary?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts [hunk 5]
```diff
@@ -1,27 +1,120 @@
...
+    private fieldUpdateSubscription: Subscription;
+
+    public languageKeys: Array<string>;
+
+    protected initializedLanguage = false;
+
+    public selectedLanguage: string;
+    public filledShown = false;
+
+    public currentValue = {};
```
[the remark this answers is not in the document]
Comment (camperko) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:27]: It's basic object with structure:
`[k:string]: string`
where key is language code and value is user given value of field for that language.

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts [hunk 6]
```diff
@@ -12,35 +12,40 @@ import {
...
+                {provide: UserResourceService, useClass: MockUserResourceService}
             ],
-            declarations: [I18nTextFieldComponent, TestWrapperComponent]
-        })
-            .compileComponents();
-    });
-
-    beforeEach(() => {
+            declarations: [I18nTextFieldComponent, TestWrapperComponent],
+            schemas: [CUSTOM_ELEMENTS_SCHEMA]
```
Comment (minop) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts:45]: why?

**Review by minop — APPROVED** [review]

File: projects/netgrif-components-core/src/assets/i18n/de.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (camperko) [reply projects/netgrif-components-core/src/assets/i18n/de.json:214]: It's in plural form, because it refers to collection of languages regardless of that collection contains only one language.

Languages in messages are currently only in English language for each one.

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts (hunk 4 above)
Comment (camperko) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts:37]: CUSTOM_ELEMENT_SCHEMA is used in each dataField test so I included it in these two tests too.
Maybe @Kovy95 knows more about it.

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts (hunk 6 above)
Comment (camperko) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts:45]: CUSTOM_ELEMENT_SCHEMA is used in each dataField test so I included it in these two tests too.
Maybe @Kovy95 knows more about it.

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kovy95) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:27]: can you make an interface for it?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts [hunk 7]
```diff
@@ -1,27 +1,120 @@
...
+            if (!(this.selectedLanguage in newValue.translations)) {
+                this.selectedLanguage = DEFAULT_LANGUAGE_CODE;
+            }
+            this.initializedLanguage = true;
+        }
+        this.currentValue = I18nField.toObject(newValue);
+        this.refreshFilledMap();
+    }
+
+    public getLanguageIcons() {
```
Comment (Kovy95) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:70]: can we add return type ?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts [hunk 8]
```diff
@@ -1,27 +1,120 @@
...
+        return this.languageIconsService.languageIcons;
+    }
+
+    public isDefaultValue(choosenLanguage: string) {
```
Comment (Kovy95) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:74]: can we add return type ?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+describe('LanguageIconsService', () => {
+  let service: LanguageIconsService;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
+    service = TestBed.inject(LanguageIconsService);
+  });
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
```
Comment (Kovy95) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.spec.ts:14]: can we add afterEach method ?

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
...012v.012l-.008.012v.012l-.004.008v.008l-.008.008zm-.152.008l.016-.008.004.004.004.008.008.004.004.008h.004V4.5h.008l.008.008h.008l.004.004h.02v-.004h.004l.004-.004V4.5l.008-.008h.004l.004-.008v-.008l.008-.004.008-.008.004-.008.02.008-.008.008-.004.012-.004.004-.008.016-.008.008-.004.008-.004.004-.008.004h-.004v.008h-.008l-.004.004h-.023l-.008-.004h-.004l-.004-.008v-.004l-.008-.004-.004-.004-.008-.004-.012-.012-.008-.016zm.051-.344v-.02.02h-.012l-.016.004-.008.004-.012.004-.023.023-.004.008-.008.012-.004.012v.012l-.004.008v.016l-.004.016v.012l.004.012v.027l.004.016v.016l.008.016v.012l.004.012.004.016.004.008.004.012.004.016.008.016.004.012.004.008.004.004-.02.012v-.008l-.008-.016-.004-.012v-.012l-.008-.008-.004-.016-.008-.023-.004-.016v-.016l-.008-.012V4.32l-.004-.012v-.035l-.004-.016v-.016l.004-.016v-.012l.004-.012v-.016l.008-.012.004-.012.016-.023.008-.008.023-.016.012-.004.016-.004h.016zm0-.02v.02zm0 0" class="sr-B"/><path d="M6.395 4.227c.066-.02.078.102.07.133-.008.043-.023.047-.055.023-.02-.027-.078-.133-.02-.156zm0 0" class="sr-D"/><path d="M6.477 4.363l-.02-.004v-.059l-.004-.004v-.016l-.004-.004V4.27l-.004-.004v-.004l-.008-.004-.004-.004-.004-.008-.008-.004h-.008l-.004-.004h
```
Comment (Kovy95) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.ts:244]: can we make domSanitizer protected ? for future extendation from another libraries

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts (hunk 4 above)
Comment (Kovy95) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts:37]: its okay, i think mostly its used in datafields to get rid of some warnings in console during tests, that you using some custom elements in html

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts [hunk 11]
```diff
@@ -1,21 +1,41 @@
...
+    TranslateLibModule, UserResourceService,
+    WrappedBoolean
+} from '@netgrif/components-core';
+import {AngularResizedEventModule} from 'angular-resize-event';
+import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
+import {HttpClientTestingModule} from '@angular/common/http/testing';
 
 describe('I18nDividerFieldComponent', () => {
     let component: I18nDividerFieldComponent;
     let fixture: ComponentFixture<TestWrapperComponent>;
```
Comment (Kovy95) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts:20]: can we add afterEach method

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss [hunk 12]
```diff
@@ -3,3 +3,130 @@
...
+    transform-origin: left;
+    color: rgba(0, 0, 0, 0.6);
+    background: transparent;
+    width: fit-content;
+    padding: 0 0.5em;
+    height: 0;
+}
+
+.invalid-form-label {
+    color: #f44336;
```
Comment (Kovy95) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss:131]: dont we had a warn color palette ? please use it if that color should be warn

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts [hunk 13]
```diff
@@ -12,35 +12,40 @@ import {
...
-import {Component} from '@angular/core';
+import {Component, CUSTOM_ELEMENTS_SCHEMA} from '@angular/core';
 import {FormControl} from '@angular/forms';
+import {AngularResizedEventModule} from 'angular-resize-event';
+import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
 
 describe('I18nTextFieldComponent', () => {
     let component: I18nTextFieldComponent;
     let fixture: ComponentFixture<TestWrapperComponent>;
+    let service: LanguageService;
```
Comment (Kovy95) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts:27]: can we add afterEach method ?

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts (hunk 6 above)
Comment (Kovy95) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.spec.ts:45]: i think its okay

File: angular.json [hunk 14]
```diff
@@ -155,7 +155,7 @@
                                 {
                                     "type": "initial",
                                     "maximumWarning": "2mb",
-                                    "maximumError": "5mb"
+                                    "maximumError": "6mb"
```
[the remark this answers is not in the document]
Comment (camperko) [reply angular.json:158]: Because of language-icon.service which contains 54 .svg icons and have 750kB which increased bundle size to about 5.2MB.

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (camperko) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:27]: Interface for basic objects is not necessary.

File: projects/netgrif-components-core/src/assets/i18n/de.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (minop) [reply projects/netgrif-components-core/src/assets/i18n/de.json:214]: Yes, but that is sadly grammatically incorrect. I don't request that both a plural and a singular form is supported, but it marginally worsens the user experience and we should at least create a Jira issue to correct it at some point in the future.

I personally see the fact that language names are not translated as a big issue and would request introducing support at least for the languages we support directly (english, german and slovak) while at the same time giving the developers an option to define the other missing translations for other languages.

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts [hunk 15]
```diff
@@ -4,15 +4,181 @@ import {Layout} from '../../models/layout';
...
+    TRANSLATION_REQUIRED = 'translationRequired',
+    TRANSLATION_ONLY = 'translationOnly',
+    REQUIRED_I18N = 'requiredI18n'
+}
+
+export const DEFAULT_LANGUAGE_CODE = 'xx';
 
 export class I18nField extends DataField<I18nFieldValue> {
 
+    public static toObject(templateValue: I18nFieldValue): any {
```
Comment (minop) [inline projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts:20]: should return at least an `object` instead of `any`. A more narrowly defined type would be preferred still.

**Review by minop — CHANGES_REQUESTED** [review]

**Review by minop — CHANGES_REQUESTED** [review]

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss [hunk 16]
```diff
@@ -3,3 +3,121 @@
     margin: 0 auto;
     width: 100%;
 }
+
+.selected-language-value {
+    outline: none;
+    padding: 9px 0;
+    width: calc(100% - 56px);
```
Comment (machacjozef) [inline projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss:10]: min-width ? IDK
maybe use clamp() ??

File: projects/netgrif-components-core/src/assets/i18n/de.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (mladoniczky) [reply projects/netgrif-components-core/src/assets/i18n/de.json:214]: An issue for language translations was created.

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts (hunk 3 above)
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:31]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts (hunk 7 above)
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts (hunk 8 above)
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.spec.ts (hunk 9 above)
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/language-icons.service.spec.ts]: yes we can 💪

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts (hunk 15 above)
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts:55]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts [hunk 17]
```diff
@@ -4,15 +4,181 @@ import {Layout} from '../../models/layout';
...
+        object[DEFAULT_LANGUAGE_CODE] = templateValue.defaultValue;
+        for (const k in templateValue.translations) {
+            if (Object.prototype.hasOwnProperty.call(templateValue.translations, k)) {
+                object[k] = templateValue.translations[k];
+            }
+        }
+        return object;
+    }
+
+    public static fromObject(templateValue: any, templateKey: string): I18nFieldValue {
```
[the remark this answers is not in the document]
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts:66]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts [hunk 18]
```diff
@@ -4,15 +4,181 @@ import {Layout} from '../../models/layout';
...
+        const i18nObject = {
+            defaultValue: templateValue[DEFAULT_LANGUAGE_CODE],
+            key: templateKey,
+            translations: {}
+        };
+        for (const k in templateValue) {
```
[the remark this answers is not in the document]
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts:72]: Done

File: projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts [hunk 19]
```diff
@@ -112,9 +116,17 @@ export abstract class AbstractI18nTextFieldComponent extends AbstractI18nErrorsC
...
     }
 
-    public getCutProperty(i18nLabel) {
-        const calculatedWidth = 'calc(0.5em + ' + i18nLabel.offsetWidth / 4 * 3 + 'px)';
-        return 'polygon(0 0, 0 100%, 100% 100%, 100% 0%, ' + calculatedWidth + ' 0, ' + calculatedWidth + ' 5%, 0.5em 5%, 0.5em 0)';
+    public getCutProperty(i18nLabel): string {
+        if (this.labelWidth !== i18nLabel.offsetWidth) {
+            this.labelWidth = i18nLabel.offsetWidth;
+            const calculatedWidth = 'calc(0.5em + ' + i18nLabel.offsetWidth / 4 * 3 + 'px)';
+            this.cutProperty = 'polygon(0 0, 0 100%, 100% 100%, 100% 0%, '
```
[the remark this answers is not in the document]
Comment (mladoniczky) [reply projects/netgrif-components-core/src/lib/data-fields/i18n-field/i18n-text-field/abstract-i18n-text-field.component.ts:123]: Done

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss (hunk 16 above)
Comment (mladoniczky) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-text-field/i18n-text-field.component.scss:10]: @machacjozef said it's ok to resolve this.

File: projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts (hunk 11 above)
Comment (mladoniczky) [reply projects/netgrif-components/src/lib/data-fields/i18n-field/i18n-divider-field/i18n-divider-field.component.spec.ts:20]: yes we can

**Review by mladoniczky — APPROVED** [review]

**Review by minop — APPROVED** [review]

**Review by machacjozef — APPROVED** [review]
